### PR TITLE
databroker server backend config

### PR DESCRIPTION
**Pull request by cuonglm**
## Summary

Allow configuring backend storage types that databroker will use.

## Related issues

**Checklist**:
- [ ] add related issues
- [x] updated docs
- [x] updated unit tests
- [ ] updated UPGRADING.md
- [x] ready for review

### CODE REVIEW

File: config/options.go [hunk 1]
```diff
@@ -482,6 +488,15 @@ func (o *Options) Validate() error {
 		log.Warn().Msg("config: cache url will be deprecated in v0.11.0")
 		o.DataBrokerURLString = o.CacheURLString
 	}
+	switch o.DataBrokerBackendStorageType {
+	case "memory":
```
Comment (desimone) [inline config/options.go:492]: fall through?

File: config/options.go [hunk 2]
```diff
@@ -225,6 +225,11 @@ type Options struct {
 	// DataBrokerURL is the routable destination of the databroker service's gRPC endpiont.
 	DataBrokerURLString string   `mapstructure:"databroker_service_url" yaml:"databroker_service_url,omitempty"`
 	DataBrokerURL       *url.URL `yaml:",omitempty"`
+	// DataBrokerBackendStorageType is the storage backend type that databroker will use.
+	// Supported type: memory, redis
+	DataBrokerBackendStorageType string `mapstructure:"databroker_backend_storage_type" yaml:"databroker_backend_storage_type,omitempty"`
```
Comment (desimone) [inline config/options.go:230]: what about DataBrokerStorageType and DataBrokerConnectionString or... something like that? DSN is a bit ambitious to me.

File: config/options.go [hunk 3]
```diff
@@ -482,6 +488,15 @@ func (o *Options) Validate() error {
 		log.Warn().Msg("config: cache url will be deprecated in v0.11.0")
 		o.DataBrokerURLString = o.CacheURLString
 	}
+	switch o.DataBrokerBackendStorageType {
+	case "memory":
+	case "redis":
```
Comment (desimone) [inline config/options.go:493]: Thoughts on adding an exported const value to the memory and redis packages so we can refer to this as variables (e.g. memory.Name / redis.Name)

File: config/options.go (hunk 2 above)
Comment (cuonglm) [reply config/options.go:230]: ok

File: config/options.go (hunk 3 above)
Comment (cuonglm) [reply config/options.go:493]: 👍

File: docs/configuration/readme.md [hunk 4]
```diff
@@ -813,6 +813,27 @@ For an example implementation, the in-memory database used by the cache service
...
+- Optional
+- Example: `redis`
+- Default: `memory`
+
+The backend storage that databroker server will use, available types: `memory`, `redis`.
+
+### Data Broker Storage Backend Connection String
+
+- Environmental Variable: `DATABROKER_BACKEND_STORAGE_CONNECTION_STRING`
+- Config File Key: `databroker_backend_storage_connection_strinng`
```
Comment (desimone) [inline docs/configuration/readme.md:830]: _strinng

File: internal/databroker/config.go [hunk 5]
```diff
@@ -8,17 +8,22 @@ var (
...
 	DefaultBTreeDegree = 8
+	// DefaultStorageType is the default storage type that Server use
+	DefaultStorageType = "memory"
 )
 
 type serverConfig struct {
 	deletePermanentlyAfter time.Duration
 	btreeDegree            int
+	storageType            string
+	storageDSN             string
```
Comment (desimone) [inline internal/databroker/config.go:19]: 👀

File: docs/configuration/readme.md (hunk 4 above)
Comment (cuonglm) [reply docs/configuration/readme.md:830]: Yay, my macbook 2019 keyboard sucks :(

File: internal/databroker/config.go (hunk 5 above)
Comment (cuonglm) [reply internal/databroker/config.go]: Fixed

File: config/options.go [hunk 6]
```diff
@@ -225,6 +227,11 @@ type Options struct {
 	// DataBrokerURL is the routable destination of the databroker service's gRPC endpiont.
 	DataBrokerURLString string   `mapstructure:"databroker_service_url" yaml:"databroker_service_url,omitempty"`
 	DataBrokerURL       *url.URL `yaml:",omitempty"`
+	// DataBrokerBackendStorageType is the storage backend type that databroker will use.
+	// Supported type: memory, redis
+	DataBrokerBackendStorageType string `mapstructure:"databroker_backend_storage_type" yaml:"databroker_backend_storage_type,omitempty"`
```
Comment (desimone) [inline config/options.go:232]: Eh, sorry to change my mind again but.. thoughts on  
`DataBrokerBackendStorageType` -> `DataBrokerStorageType`
`DataBrokerStorageConnectionString` -> `DataBrokerStroageConnectionString`

File: config/options.go [hunk 7]
```diff
@@ -225,6 +227,11 @@ type Options struct {
...
+	// DataBrokerBackendStorageConnectionString is the data source name for storage backend.
+	DataBrokerBackendStorageConnectionString string `mapstructure:"databroker_backend_storage_connection_string" yaml:"databroker_backend_storage_connection_string,omitempty"`
```
Comment (desimone) [inline config/options.go:234]: @calebdoxsey will these fields need to be updated in the settings pb definitions?

File: config/options.go (hunk 6 above)
Comment (cuonglm) [reply config/options.go:232]: 👍

File: config/options.go (hunk 7 above)
Comment (calebdoxsey) [reply config/options.go]: Yes if we change the settings in pomerium we also have to change them in the dashboard.

Comment (cuonglm) [reply config/options.go]: Not sure it's necessary, I guess we only store policies and routes config in databroker?

**Pull request merged**